src/app/components/category-products/category-products.component.ts

```diff
@@ -8,4 +21,160 @@ import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/
+
+  readonly model$: Observable<CategoryProductsQueryModel> = combineLatest([
+    this._filters$,
+    this._categoryService.getAll(),
+    this._productService.getAll()
+  ]).pipe(
+    map(
+      (
+        [filters, categories, products]: [CategoryProductsFiltersQueryModel, CategoryModel[], ProductModel[]]
+      ) => this._mapQueryModel(filters, categories, products)
+    )
+  );
+
+  constructor(private _categoryService: CategoryService, private _activatedRoute: ActivatedRoute, private _productService: ProductService, private _router: Router) {
+  }
+
+  public onSortingChange(target: EventTarget | null): void {
+    if (target === null) return;
+
+    const value = (target as HTMLSelectElement).value;
+
+    this._navigate({ sort: value });
+  }
+
+  public onLimitChange(limit: number): void {
+    this._navigate({ limit: limit });
+  }
+
+  public onPageChange(page: number): void {
+    this._navigate({ page: page });
+  }
+
+  private _setPages(filters: CategoryProductsFiltersQueryModel, productsCount: number): void {
+    const maxPage = Math.ceil(productsCount / filters.limit);
+
+    const pages = Array.from(Array(maxPage).keys()).map(x => ++x);
+
+    this._pagesSubject.next(pages);
+
+    if (filters.page > maxPage) {
+      this._navigate({ page: maxPage })
+    }
+  }
+
+  private _navigate(queryParams: Object) {
+    this._router.navigate(
+      [],
+      {
+        relativeTo: this._activatedRoute,
+        queryParams: queryParams,
+        queryParamsHandling: 'merge'
+      }
+    )
+  }
+
+  private _filterProducts(filters: CategoryProductsFiltersQueryModel, products: ProductModel[]): ProductModel[] {
+    return products.filter(product => product.categoryId === filters.categoryId);
+  }
+
+  private _sortProducts(filters: CategoryProductsFiltersQueryModel, products: ProductModel[]): ProductModel[] {
+    const sortFunctionsMap = new Map([
+      [ProductSorting.featureValueDesc, (a: ProductModel, b: ProductModel) => b.featureValue - a.featureValue],
+      [ProductSorting.priceAsc, (a: ProductModel, b: ProductModel) => a.price - b.price],
+      [ProductSorting.priceDesc, (a: ProductModel, b: ProductModel) => b.price - a.price],
+      [ProductSorting.ratingValueDesc, (a: ProductModel, b: ProductModel) => b.ratingValue - a.ratingValue],
+    ]);
+
+    const sortFunction = sortFunctionsMap.get(filters.sort) ?? sortFunctionsMap.get(ProductSorting.featureValueDesc);
+
+    const sortedProducts = products.sort(sortFunction);
+
+    return sortedProducts;
+  }
+
+  private _paginateProducts(filters: CategoryProductsFiltersQueryModel, products: ProductModel[]): ProductModel[] {
+    return products.slice(
+      (filters.page - 1) * filters.limit,
+      filters.page * filters.limit
+    );
+  }
+
+  private _mapQueryModel(
+    filters: CategoryProductsFiltersQueryModel,
+    categories: CategoryModel[],
+    products: ProductModel[]
+  ): CategoryProductsQueryModel {
+    products = this._filterProducts(filters, products);
```

Comment:
I dont like mutating same variable, you can just make it all in 1 chain of function, seems cleaner to me